ctapipe/image/extractor.py

```diff
@@ -104,16 +106,69 @@ def neighbor_average_waveform(waveforms, neighbors, lwt):
     return sum_ / n
 
 
-class ChargeExtractor(Component):
+def extract_pulse_time_weighted_average(waveforms):
+    """
+    Use the weighted average of the waveforms to extract the time of the pulse
+    in each pixel
+
+    Parameters
+    ----------
+    waveforms : ndarray
+        Waveforms stored in a numpy array.
+        Shape: (n_chan, n_pix, n_samples)
+
+    Returns
+    -------
+    pulse_time : ndarray
+        Floating point pulse time in each pixel
+        Shape: (n_chan, n_pix)
+
+    """
+    samples_i = np.indices(waveforms.shape)[2]
+    pulse_time = np.average(samples_i, weights=waveforms, axis=2)
+    pulse_time[pulse_time < 0] = 0
+    pulse_time[pulse_time > waveforms.shape[2]] = waveforms.shape[2]
```

Comment:
I'm not sure ... is it possible to have so funny weights, that the pulse_time can be outside of the range `[0...waveforms.shape[2]]`? ... I think not... unless ... of course .. when the weights are negative, then the average can be pulled out of range, but I think this should be avoided. 

So I played a little bit around with this kind of average ... I used this kind of signal:
![image](https://user-images.githubusercontent.com/8200858/55553036-62651c00-56df-11e9-9d68-f6b903294ea1.png)

As one can see the actual pulse start time and the result of this `np.average(sample_i, weights=waveforms)` is clearly correlated as expected. 

But as soon as the baseline moves around problems occur:
![image](https://user-images.githubusercontent.com/8200858/55553639-026f7500-56e1-11e9-8d87-89c329b28582.png)

![image](https://user-images.githubusercontent.com/8200858/55553649-0b604680-56e1-11e9-898b-400cd1f535e5.png)

![image](https://user-images.githubusercontent.com/8200858/55553658-11562780-56e1-11e9-93ed-3c9ad72be3d2.png)

![image](https://user-images.githubusercontent.com/8200858/55553664-16b37200-56e1-11e9-8f59-d9ef42cd7cc0.png)

So I think there needs some care to be taken with respect to these weights. And one should not simply say ... ah ... when the pulse_time is negative .. then we set it to zero ... I think when the pulse_time is out of range, this simply is a strong hint, that something was not good about the range of these weights.

Reply:
Ah sorry .. I just see, I labelled my axis wrongly: What I called `average_time` is actually called `pulse_time`. 

So `dneise_average_time = pulse_time = np.average(samples_i, weights=waveform)`

Reply:
Yeah you are absolutely right. This implementation for obtaining pulse time is very simple, but not very robust. The limit I added for the pulse time to be within the waveform readout was done to make it easier to plot this. Thinking about it now, it could be better to set such unrealistic pulse times to nan (what do you think?).

Or do you suggest only use samples with positive weights in the calculation?

I am inclined to suggest that contributions are made for better pulse time algorithms in future pull requests, which can replace the current function used in the ImageExtractors. Unless you are adamantly against accepting this implementation temporarily.

Reply:
Maybe not cut at all then. The result of this implementation is (even in the best case) only correlated with and not identical to the arrival time (50% of rising edge) .. so it is difficult to say what the allowed range is ... I think .. so maybe avoiding a cut and leave the cut to the user of this implementation is ok.

Reply:
No no .. simple implementations are good .. just this cut was puzzling me.

Reply:
Without the cut, it made plotting tools (such as the ctapipe-event-viewer) useless for checking the pulse time, due to the z axis range. I would rather not impose a limit inside this plotting tool, as then it is a misleading visualisation of the data. 

I preferred having these cuts in the algorithm because it's not physical to have a pulse time outside the window.

I would prefer to improve the implementation rather than outright remove the cuts, should you request it.

Would setting it to nan be more acceptable?
How do your plots look if you only consider positive samples in the weighted average?

Reply:
When cutting away the negative weights the correlation looks a bit better, but mind the difference in x and y axis ranges .. y goes from (15, 70) while x is from (0, 50). I think this is an effect of the pulse shape... and this is independent of the offset or the noise ...

Reply:
![image](https://user-images.githubusercontent.com/8200858/55558396-4a47c980-56ec-11e9-9a5c-d6733d9d7de4.png)

Reply:
Okay, that seems somewhat better, thanks. However, with this implementation we would have to handle the case where there are no positive samples, which would involve setting the pulse_time to some value such as zero.

Could we reserve consideration of better pulse_time algorithms for a different PR?

Reply:
I have added it as a task in #1026

Reply:
Indeed if you set out of range results to nan that might be better, since the user knows there was indeed a cut.

Reply:
> Could we reserve consideration of better pulse_time algorithms for a different PR?

totally. You asked me do produce these plots.